Add KEP #257: Leader Only SubGroup (#402)
kubernetes-sigs/lws: keps/257-Subgroup-leader-only/README.md

motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

There are cases where the resource requirements for the leader are different from the ones by the worker, while still being able to guarantee exclusive placement
on the workers. Having a leaderOnly subgroup means that the placement of the leader will be independent of the workers, and by adding the `subgroup-exclusive-topology`, the 
workers can be guaranteed to be scheduled on the same topology.

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

Add an option to create a LeaderOnly subgroup, while still being able to create subgroups on the other workers. Therefore, it should be possible to do the following split:
(leader), (worker-1, worker-2), (worker-3, worker-4). If the desired effect is to have one subgroup for the leader, and one subgroup for the worker, subgroupSize should be set 
to the number of workers (so size - 1). 

### Non-Goals

> This KEP assumes that the leader will not request TPU resources, and thus, subGroupSize will always be assumed to be an odd number. Moreover, no TPU environment variables will be 

Author: Both TPU and GPU are supported for exclusive placement. I'll rephrase to be accelerator agnostic

Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

A new API field will be added under the SubGroupPolicy type. This API field will determine what type 
of subgroup is desired. The default is LeaderWorker, which is the existing implementation of subgroup. `SubGroupSize`
will continue to be a required field, while `Type` will be optional. 

```golang
type SubGroupPolicy struct {

	// +kubebuilder:validation:Enum={LeaderWorker,LeaderOnly}
	// +kubebuilder:default=LeaderWorker
	Type SubGroupPolicyType `json:"subGroupPolicyType,omitempty"`

    SubGroupSize *int32 `json:"subGroupSize,omitempty"`
}

type SubGroupPolicyType string

const (
	SubGroupPolicyLeaderWorker SubGroupPolicyType = "LeaderWorker"

> 	SubGroupPolicyLeaderOnly SubGroupPolicyType = "LeaderOnly"

Author: I'm fine with `LeaderAlone`, but `All` sounds a bit ambiguous to me. `LeaderAlone` vs `LeaderWorker` maybe? Or something that specifies that the leader and a worker pod will be part of the same subgroup

)
```

A new annotation will be created to determine whether or not the subgroup type is LeaderOnly

* `leaderworkerset.sigs.k8s.io/subgroup-policy-type`

> In order to keep backwards compatability, it will only be added if the type is `LeaderOnly`.

Reviewer: I think even we add the annotation for both types is still backward compatibility. No behavior change.

Author: The scenario that I see is that there is an existing deployment using `SubGroupPolicy` which then upgrades the LWS controller. After the upgrade, it will inject the annotation on the PodSpec, which will then trigger a rolling update since the Pod has changed.

### SubGroupIndex and TPU Environment Injection
SubGroupIndex is a label that we use to generate the hash that is used for exclusive affinity

```golang
subGroupUniqueKey := genGroupUniqueKey(leaderName, subGroupIndexKey)
pod.Labels[leaderworkerset.SubGroupUniqueHashLabelKey] = subGroupUniqueKey
if subEpKey, foundSubEpKey := pod.Annotations[leaderworkerset.SubGroupExclusiveKeyAnnotationKey]; foundSubEpKey {
    SetExclusiveAffinities(pod, subGroupUniqueKey, subEpKey, leaderworkerset.SubGroupUniqueHashLabelKey)
}
```
and what we use to determine what pod names to include when injecting the TPU Host names 

```golang
start := subGroupSize*subGroupIndex + 1
end := subGroupSize * (subGroupIndex + 1)
for i := start; i <= end; i++ {
    hostnames = append(hostnames, fmt.Sprintf("%s-%d.%s", leaderName, i, pod.Spec.Subdomain))
}
```

> Because the leader now occupies a subgroup, and `SubGroupSize` will not always be one, the way we calculate the SubGroupIndex needs to be modified. When generating 

Reviewer: One argue here is should we mark the leader a sub group? What's the Cons if we don't do this.

Author: That's a good point, hadn't considered not making the leader part of a subgroup at all. Doing it this way simplifies the rest of the changes (though the TPU environment injection will still need some adjustment). Will test it out and add to the `Alternatives` which ever version we don't implement

- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

## Summary

<!--
This section is incredibly important for producing high-quality, user-focused
documentation such as release notes or a development roadmap. It should be
possible to collect this information before implementation begins, in order to
avoid requiring implementors to split their attention between writing release
notes and implementing the feature itself. KEP editors and SIG Docs
should help to ensure that the tone and content of the `Summary` section is
useful for a wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->
This KEP aims to extend the SubGroup feature by adding an API field that defines the type of Subgroup that will be created. The existing version

> will become "LeaderWorker", and a new type will be added to create a SubGroup that only includes the leader.

Reviewer: There's no subgroup for leader pod now.

Author: Right, updated this and all other references to leader being its own subgroup

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

There are cases where the resource requirements for the leader are different from the ones by the worker, while still being able to guarantee exclusive placement

> on the workers. Having a leaderOnly subgroup means that the placement of the leader will be independent of the workers, and by adding the `subgroup-exclusive-topology`, the 

Reviewer: There's no leaderOnly type as well.

workers can be guaranteed to be scheduled on the same topology.

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

> Add an option to create a LeaderOnly subgroup, while still being able to create subgroups on the other workers. Therefore, it should be possible to do the following split:

Reviewer: ditto

motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

There are cases where the resource requirements for the leader are different from the ones by the worker, while still being able to guarantee exclusive placement
on the workers. Having a leaderOnly subgroup means that the placement of the leader will be independent of the workers, and by adding the `subgroup-exclusive-topology`, the 
workers can be guaranteed to be scheduled on the same topology.

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

Add an option to create a LeaderOnly subgroup, while still being able to create subgroups on the other workers. Therefore, it should be possible to do the following split:
(leader), (worker-1, worker-2), (worker-3, worker-4). If the desired effect is to have one subgroup for the leader, and one subgroup for the worker, subgroupSize should be set 
to the number of workers (so size - 1). 

### Non-Goals

> This KEP assumes that the leader will not request TPU resources, and thus, subGroupSize will always be assumed to be an odd number. Moreover, no TPU environment variables will be 

Reviewer 2: pls rephrase to be accelerator agnostic, this applies to both

injected on the leader.

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

> As a user, I should be able to create a subGroup that only includes the leader in order to be able to schedule the leader on a different topology while 

Reviewer 2: ```suggestion
As a user, I should be able to create a subGroup that only includes the leader in order to be able to schedule the leader with different resource requirements
```

This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

A new API field will be added under the SubGroupPolicy type. This API field will determine what type 
of subgroup is desired. The default is LeaderWorker, which is the existing implementation of subgroup. `SubGroupSize`
will continue to be a required field, while `Type` will be optional. 

```golang
type SubGroupPolicy struct {

	// +kubebuilder:validation:Enum={LeaderWorker,LeaderAlone}
	// +kubebuilder:default=LeaderWorker

> 	Type SubGroupPolicyType `json:"subGroupPolicyType,omitempty"`

Reviewer 2: fix indentation pls

    SubGroupSize *int32 `json:"subGroupSize,omitempty"`
}

type SubGroupPolicyType string

const (
	SubGroupPolicyLeaderWorker SubGroupPolicyType = "LeaderWorker"

	SubGroupPolicyLeaderAlone SubGroupPolicyType = "LeaderAlone"
)
```

A new annotation will be created to determine whether or not the subgroup type is LeaderAlone

* `leaderworkerset.sigs.k8s.io/subgroup-policy-type`

This annotation will only be injected in the leader pod.

In order to keep backwards compatability, it will only be added if the type is `LeaderAlone`.

### Subgroup Creation 

> Implementation wise, the only change needed is to not add the SubGroup labels on the leader if the SubGroupType is LeaderAlone. Effectively, this means 

Reviewer 2: The indexing within the group will be different now. Currently we don't define an index within the group, we do it only for tpuWorkerId, and this should be calculated differently now to exclude the leader: https://github.com/kubernetes-sigs/lws/blob/main/pkg/utils/accelerators/tpu.go#L113, right?

Author: We already do this when the leader doesn't request TPU resources https://github.com/kubernetes-sigs/lws/blob/main/pkg/utils/accelerators/tpu.go#L115-L117

> that the leader is not part of a subgroup at all. The only point of a pod being in a subgroup is to guarantee exclusive placement with the other pods 

Reviewer: > the leader is not part of a subgroup at all

LeaderAlone is based on the original assumption that leader pod is a separate group, if not, maybe we should rename it to like `leaderExcluded`, in the future, leaderAlone can still be reuse if leader pod should be the 0-index subgroup.

Author: Makes sense to me, will change it to `leaderExcluded`. Does `leaderWorker` then sound good to you for the default setup?

Reviewer: LGTM
